### DISCUSSION TITLE
How to use dynamic routing and dynamic data, without returning millions of ids for getStaticPaths, probably we need to use getServerSideProps

### DISCUSSION THREAD

**Discussion by andreidiaconescu** [vercel/next.js#36068] [Help]
Hello,
the tutorial seems really nice (https://nextjs.org/learn/basics/data-fetching/request-time and https://nextjs.org/learn/basics/dynamic-routes/dynamic-routes-details), 
but, for me, it does not explain a case that seems really most common:  How to use dynamic routing and dynamic data, without static rendering ( without returning millions of ids for getStaticPaths);
For example we have in the DB millions of Articles, and we want to display tham by url that contains their id.
Probably we need to use  is dynamic routing together with getServerSideProps.
Any examples on this ?

thank you,

**Comment by icyJoseph** [comment]
Say for example that you have to render `/pokemon/[id]`, seeing that there's 800+ Pokemon's (last time I checked), that'd be a really long build time. 

We can do the following (ISR):

1. Return 3 (absolutely arbitrary number) paths from `getStaticPaths`, for Pokemon's 1, 4 and 7.
2. Include the `fallback: true` or `fallback: "blocking"`, props in the `getStaticPaths` response
3. Make sure `getStaticProps` handles `notFound` or `redirect` cases for malformed paths, or non existing data (apply revalidate to `notFound` as well)
4. Deploy your app
5. A request arrives for Pokemon 300, which you have not generated in step 1
6. ISR kicks in because of `fallback` being set in step 2
7. A page for 300 is generated and from now on, everyone arriving to it will immediately see it
8. If you render a page with a bunch of `pokemon/[id]` hrefs using `next/link`, these are also generated 

You can take a look at this demo I built for a MeetUp on my city https://github.com/icyJoseph/next-js-react-gbg, which does exactly the above, at this route https://github.com/icyJoseph/next-js-react-gbg/blob/main/demo-app/src/pages/pokemon/%5Bid%5D.tsx.

In your particular case, I'd return the top paths, by some metric (for example: popularity, or amount of revenue per page), and let the fallback true option handle the rest.

When you use `fallback: true` Client Side SPA like transitions, will need to check `useRouter().isFallback`, and if you use `fallback: "blocking"` then the server waits until the page is ISR generated. Crawlers and bots, will never see fallback UI.

Also, `revalidate` helps you to keep these pages updated as people navigate to them.

---

`getServerSideProps` could also do this if you combine it with a cache layer, or a CDN kind of solution. Please refer to https://nextjs.org/docs/going-to-production#caching, and do read the entire thing. 

In short you have to set `cache-control` on the response from `getServerSideProps`, and then your caching layer should make sure to replicate this to all users (otherwise it'd scoped only to a user).

> Note: Your deployment provider must support caching for dynamic responses. If you are self-hosting, you will need to add this logic yourself using a key/value store like Redis. If you are using Vercel, [Edge Caching works without configuration](https://vercel.com/docs/edge-network/caching).
